remote_syslog.go

```diff
@@ -17,111 +13,57 @@ import (
-			logger.Packets <- syslog.Packet{
-				Severity: severity,
-				Facility: facility,
-				Time:     time.Now(),
-				Hostname: logger.ClientHostname,
-				Tag:      path.Base(file),
-				Message:  line.Text,
-			}
-			log.Tracef("Forwarding: %s", line.Text)
-		} else {
-			log.Tracef("Not Forwarding: %s", line.Text)
-		}
-
+	// run in the background if asked to
+	if config.Daemonize {
+		utils.Daemonize(config.DebugLogFile, config.PidFile)
 	}
-
-	log.Errorf("Tail worker executed abnormally")
-}
-
-// Tails files speficied in the globs and re-evaluates the globs
-// at the specified interval
-func tailFiles(globs []string, excludedFiles []*regexp.Regexp, excludePatterns []*regexp.Regexp, interval RefreshInterval, logger *syslog.Logger, severity syslog.Priority, facility syslog.Priority, poll bool) {
+	// set up logging
+	loggo.ConfigureLoggers(config.LogLevels)
+	// connect to remote syslog
+	host := config.DestHost
+	raddr := net.JoinHostPort(host, strconv.Itoa(config.DestPort))
+	log.Infof("Connecting to %s over %s", raddr, config.Protocol)
+	rootcas := config.RootCAs
+	logger, err := syslog.Dial(
+		config.Hostname,
+		config.Protocol,
+		raddr,
+		rootcas,
+	)
+	if err != nil {
+		log.Criticalf("Cannot connect to server: %v", err)
+		os.Exit(1)
+	}
+	// tail files
 	wr := NewWorkerRegistry()
-	log.Debugf("Evaluating globs every %s", interval.Duration)
-	logMissingFiles := true
+	log.Debugf("Evaluating globs every %s", config.RefreshInterval)
+	warn := true
 	for {
-		globFiles(globs, excludedFiles, excludePatterns, logger, &wr, logMissingFiles, severity, facility, poll)
-		time.Sleep(interval.Duration)
-		logMissingFiles = false
-	}
-}
-
-//
-func globFiles(globs []string, excludedFiles []*regexp.Regexp, excludePatterns []*regexp.Regexp, logger *syslog.Logger, wr *WorkerRegistry, logMissingFiles bool, severity syslog.Priority, facility syslog.Priority, poll bool) {
-	log.Debugf("Evaluating file globs")
-	for _, glob := range globs {
-
-		files, err := filepath.Glob(utils.ResolvePath(glob))
-
-		if err != nil {
-			log.Errorf("Failed to glob %s: %s", glob, err)
-		} else if files == nil && logMissingFiles {
-			log.Errorf("Cannot forward %s, it may not exist", glob)
-		}
-
-		for _, file := range files {
-			switch {
-			case wr.Exists(file):
-				log.Debugf("Skipping %s because it is already running", file)
-			case matchExps(file, excludedFiles):
-				log.Debugf("Skipping %s because it is excluded by regular expression", file)
-			default:
+		files, err := glob(config.Files, config.ExcludeFiles, wr, warn)
+		if err == nil {
+			for file, _ := range files {
 				log.Infof("Forwarding %s", file)
-				go tailOne(file, excludePatterns, logger, wr, severity, facility, poll)
+				go tailone(
+					file,
+					config.ExcludePatterns,
+					config.Severity,
+					config.Facility,
+					config.Poll,
+					logger,
+					wr,
+				)
+				for err = range logger.Errors() {
+					log.Errorf("Syslog error: %v", err)
+				}
```

Comment:
As this is iterating over a channel that won't be closed until r_s2 exits, would we get stuck here and never continue the outer loop over `files`. Also, should this check begin before `tailone` starts executing in case errors are generated early on?